tracee-ebpf/tracee/events_pipeline.go

```diff
@@ -39,25 +40,39 @@ type context struct {
 	_        [3]byte //padding
 }
 
-func (t *Tracee) processEvents(done <-chan struct{}) error {
+// handleEvents is a high-level function that starts all operations related to events processing
+func (t *Tracee) handleEvents(done <-chan struct{}) {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+
+	eventsChan := make(chan *external.Event)
+	go t.decodeEvents(ctx, eventsChan)
+	go t.processEvents(ctx, eventsChan)
```

Comment:
when you wanted to remove the pipeline you had a good case that it's too much for what we needed. Why do you want to bring it back now? more precisely, how goes calling the decode and process functions as goroutines help?

Reply:
When we removed the pipeline, it looked differently.
The code was much more complex (argTag and other stuff, which were removed since), and its original usage was not relevant anymore (as we added tracee-rules).
I think that the split introduced in this PR is much more readable, and makes a lot of sense.
The decode stage is responsible for creating the external.Event struct, which is the minimal struct to work with for any future work we want to do with it. 
For example, sorting events doesn't have to wait for their processing, and can reorder the events right after it has been decoded. In addition, such sorting will have to take place as another stage in the pipeline, as it needs to keep events until sync is achieved between the different cpu queues (#1103), so another channel is required, and we get back to pipeline anyways.

Reply:
> the split introduced in this PR is much more readable

I agree the split is good, but the question is why a pipeline. If you want to extract a function, that's always good, but if you want to make the function a go routine that's something else.

Reply:
> I think that the split introduced in this PR is much more readable, and makes a lot of sense.

I agree the split is good (the extracted functions). I just question the use of goroutines. 
If the motivation is supposed performance improvement thanks for parallelism - that's might be counter-productive, as @danielpacak had discovered when investigating tracee-rules pipeline. 
If the motivation is related to event sorting - ok, but I'm not clear how this pipelining this part helps

Reply:
Itay has a point.
The processing part doesn't have to be in its own go routine.
The sorting will have to be parallel, but the processing before and after it could be all one go routine (one for before and one for after) and by that reduce the performance problems of channels and context switching and avoid potential synchronization bugs.
The only thing that is bothering me is that we want to read as fast as we can from the data channel so the perf-buffers won't get full and result event loss. So for this reason, we might want the events decoding to be separated from other logics by different go routines.
Also, the pipeline design pattern make it very easy to expand the stages and add new features to events processing. So there are advantages.

Reply:
> we want to read as fast as we can from the data channel so the perf-buffers won't get full and result event loss. So for this reason, we might want the events decoding to be separated from other logics by different go routines.

Agree. I think this was the motivation for the original pipeline. but since the connecting channels are unbuffered, I don't this it'll make any difference. the reading gorouting would just block until the downstream goroutines would proceed

Reply:
> the pipeline design pattern make it very easy to expand the stages and add new features to events processing. So there are advantages.

definitely. but for that to happen I think we need to address https://github.com/aquasecurity/tracee/pull/1323#discussion_r779386915

Reply:
> I agree the split is good (the extracted functions). I just question the use of goroutines.
> If the motivation is supposed performance improvement thanks for parallelism - that's might be counter-productive, as @danielpacak had discovered when investigating tracee-rules pipeline.
> If the motivation is related to event sorting - ok, but I'm not clear how this pipelining this part helps

I think you should have a look at #1103 and then you will have a better understanding for why this is required (and maybe you will have ideas for eliminating the need for the pipeline, who knows...).
About the performance part - I don't know what @danielpacak found in tracee-rules, but measuring the performance gain with a pipeline might be tricky. The time that a single event spends in the pipeline is longer than without the pipeline as the amount of computation is more or less the same, but an extra channel is added, which adds to the time the event stays in the pipeline. Yet, the throughput should be better, as things happen in parrallel, so more events can be processed in the same time window. Another thing to take into account, which might also be counter intuitive, is that the cpu usage might look higher with the pipeline when there is a high events load. The reason for this is that now two (or more) go routines are occupied with processing the events (instead of one without the pipeline), and if events keep coming in a high frequency, cpu load will be higher. Yet, the throughput should be higher, which is a good thing.

Reply:
IIRC, at the time when we looked into tracee-rules CPU utilisation the requirement was quite opposite. Our goal was to lower CPU from spikes even if we sacrifice throughput.

Reply:
But is this what we really want?
Doing things in parallel will always mean having a greater cpu utilization for getting higher throughput. So if optimizing for no cpu spikes, the reason says that you need to kill all the parallelism. This makes no sense to me.

Reply:
Isn't this a need for "tuning settings" feature ? 

Can't we implement tuning (though config options) to the amount of parallelism we would like to have ? 

Depth of channel buffering, amount of go routines we would like to run in parallel... We could even set the time window size we would like the events buffer to sort events.

Of course we can have a good "generic" number... But tuning things like that will permit us to maximize performance in real big machines, for example, where we can dedicate "tracee" to specific CPUs and considering amount of threads running.

This type of technique is used to maximize throughput in big cloud servers using DPDK and QEMU, for example. I'm just thinking that, with tunables, we would be able to pin tracee to specific CPUs dedicated to security in a big cloud server (for example), and other stuff (we could even recommend final users to use NOHZ, or disable RCU callbacks, so there are no uneeded context switches if we want a huge performance gain).

Just a thought that could help tracee being deployed and customized in big environments.

Reply:
@yanivagman  TBH at that time the requirement to use as less CPU as possible did not make sense to me as well.

My conclusion was similar to what @rafaeldtinoco mentioned, i.e. that maybe we could control the number of goroutines used by tracee based on user config or number of available CPUs (e.g. respecting GOMAXPROCS env).

Reply:
Throughput and cpu utilization are two different metrics to measure performance.
Deciding on the right metric to measure performance depends on the application and business needs.
For a server application, the throughtput might be the more important metric to measure.
However, for tracee we should probably balance between those two.
Having a higher throughput will enable tracee to handle bigger loads and peaks without loosing events (as the buffers have a limited amount of space) and respond quickly.
On the other hand, tracee is not the main business logic in a node, because it is used to monitor other applications in the system. This means that it competes with other applications on cpu resources, and cpu utilization then becomes important.
We can improve both of these metrics if we remove or optimize logic in hotspots, but having paralellism will benefit the throughput metric, while degrading cpu utilization.
Having the ability to balance these two metric with some kind of tuning is definitely a desired feature that we should consider in the future. For this PR, I think it is not a bad thing to have some parallelism and gain some throughput, even if that might affect cpu utilization, which can probably be solved with other measures, such as limiting the cpu resources with cgroups.

Reply:
I have read the discussion and I correct me if I'm wrong but the question is still unanswered - what do we gain from calling those functions as goroutines?
as long as the channels are unbuffered we don't gain throughput or performance.
Also, just to address some of the questions, we should assume Tracee (not tracee-ebpf) is allocated 1 core

Reply:
> I have read the discussion and I correct me if I'm wrong but the question is still unanswered - what do we gain from calling those functions as goroutines? as long as the channels are unbuffered we don't gain throughput or performance. Also, just to address some of the questions, we should assume Tracee (not tracee-ebpf) is allocated 1 core

I don't think it's correct to say that with unbuffered channels we won't get performance gain. You have two go routines that work in parallel - then you get more throughput. Using buffered channel in a pipeline won't help, as the throughput will still be determined by the "weakest link" in the pipeline.
Other than that and like I wrote before - anyways we are going to need a pipeline for the sorting feature. I don't think there is a way around that.